### DISCUSSION TITLE
Planned Pulseq fromat extensions

### DISCUSSION THREAD

**Discussion by m-a-x-i-m-z** [pulseq/pulseq#68] [Ideas]
Dear all, 

we intend to modify the Pulseq file format slightly within the next month or so (e.g. before mid Dec 2024). The planned changes include: 

1. adding 'RF center' and 'intended use' fields to the RF pulse object to enable accurate and robust k-space calculation 
2. adding 'first' and 'last' properties to the GRAD objects
3. adding phase modulation vector to the ADC object to enable e.g. off-center non-Cartesian imaging 

We have a whole host of further ideas, which we however are likely to keep for the subsequent format updates. These include: 

- RF pulse extensions for parallel transmit and/or RF shimming
- ADC masking and oversampling 
- GRAD oversampling 

Please feel free to use this thread to propose further ideas or vote the existing proposals up or down. 

--M.

**Comment by btasdelen** [comment]
Hi Maxim. I think these are great improvements.

This might be a significant undertaking, but one feature very much desired by non-Cartesian people is the extension for rotating readout by rotation matrices. Rotating readouts and adding them as arbitrary waveforms 1) limits how large the sequence is due to memory issues, 2) generates huge file sizes.

Also, a minor suggestion: Can we make setting the image resolution by seq definition an option?

**Reply by zhengliuX** [reply]
Hi, just out of curiosity, is the goal of setting image resolution in definition field to display the right value on the UI when scanning?

**Reply by btasdelen** [reply]
Yes, but also to correctly fill MDH headers, so that a generic reconstruction knows the matrix size.

**Reply by zhengliuX** [reply]
I am interested in this. On our platform(UIH), I added a field called Matrix, to define the image matrix size in RO, PE and Slice/PE_2nd. In this way, I could integrate the Pulseq sequence into the vendor recon pipeline(also with some other parameters, like the LastPeIndex and LastParIndex). But apart from that, having the info in the definition field is helpful, especially for some people dont know the meta info of the .seq file.

**Reply by fzimmermann89** [reply]
Hi,

I would also be interested in setting the `phaseEncodingLines`  to a custom value.
If I am not missing anything, this would be a requirement to support openrecon reconstruction of non-cartesian sequences on Siemens scanners (due to the openrecon ice expecting an image with y=phaseEncodingLines). 
Currently, the only work-around is using one of the other counters for the lines and relying on the phaseEncodingLines=baseresolution fallback case. 
@m-a-x-i-m-z would you consider adding this via a definition to the format and siemens interpreter?

Cheers
Felix

**Reply by m-a-x-i-m-z** [reply]
@fzimmermann89: yes the best way would be to add to to the definitions that are mapped by the interpreter to SeqExpo/YAPS. 

By the Siemens interpreter supports the following definitions: DominantGradientSwitchingFrequency, kSpaceCenterLine, kSpaceCenterPartition, kSpaceCenterSample, Name, MaxAdcSegmentLength, OrientationMapping, PhaseResolution, ReadoutOversamplingFactor, ReceiverGainHigh, SliceGap, SlicePositions, SliceThickness, Slice_Thickness, TargetGriddedSamples, TE, TotalDuration, TR, TrapezoidGriddingParameters. Their meaning and interpretation is described in the table in Section 4.6 of the interpreter package documentation. Only few of them are documented in the format specification document. We should probably change that.

@all: I think we should create a separate discussion to collect such mappings or which ever required metadata information for the future versions.

**Reply by zhengliuX** [reply]
That would be nice. I believe for the general infomation, such as image matrix, or PE center, we could have some recommended definitions for different vendors' interpreter to refer to.

**Comment by namgyunlee** [comment]
Hi Maxim. Thank you for your initiation.
What @btasdelen suggested is indeed a practical issue that makes the development of some methods difficult such as
spiral-based [bSTAR ](https://onlinelibrary.wiley.com/doi/full/10.1002/mrm.29757)(free-breathing lung imaging using bSSFP) and [MR fingerprinting based on spiral projection imaging](https://onlinelibrary.wiley.com/doi/10.1002/mrm.29194).
In these applications, spiral trajectories are rotated in various ways to maximize incoherence in undersampling artifacts, often requiring more than 100,000 rotated spiral arms. This is very difficult to achieve in the current file format without disrupting signal evolution.

**Reply by m-a-x-i-m-z** [reply]
Dear all, 

as things stand right now and with the current infrastructure in place we see it really impractical to add support for block-wise rotation matrices. The problem is that the additional overhead of needing to rotate everything every time will impact the performance of all other sequences extremely heavily. 

What we are thinking in the long term is a concept of sub-sequence(s), which will be individually rotatable. But these are quite distant plans. 

Another really spontaneous idea, which came to my mind as I was writing this response is to introduce an ability of rotating individual gradients, which should solve the problem of having to repeatedly sore the shapes, which are just rotated copies of each other... Need to think this to the end though, because it may introduce conflicts with gradients on the fixed axes... 

But once again, all these are incompatible with the December timeline... 

--Maxim

**Reply by WSHoge** [reply]
Hey all, but especially @btasdelen.  The ability to rotate in 3D a single readout is a high priority for a project I'm currently working on.  So, following an idea listed elsewhere (https://github.com/imr-framework/pypulseq/discussions/184) I've gone ahead and modified the matlab generator to add rotation matrices as an extension to the gradient block.  I'll be tracking my progress on a pulseq fork (https://github.com/WSHoge/pulseq/tree/rotext) in the hopes that at some point it will be useful to others. 

I'm happy to discuss and welcome feedback. I'll note that I'm currently testing pulseq on both Siemens and GE, so I'm hopeful a harmonized solution for at least those two platforms will be possible.

**Reply by m-a-x-i-m-z** [reply]
Hi Scott, 

why don't you star a new thread and describe what has readily been done and tested in the existing forks. I'm mostly interested to hear different opinions on the following topics: 

* how to describe/store rotations (options are axis/angle; quaternion, Euler angles or rotation matrix)
* at which stage to apply transformation to make sure that gradient consistency checks, plotting and k-space calculation work 
* additionally, we have recently implemented FOV positioning functions. These will also be challenged by the rotation extension. 

Best regards, 
Maxim

**Comment by m-a-x-i-m-z** [comment]
BTW, the current Dev branch of the Pulseq-Matlab environment has a new function mr.rotate3D() which rotates gradients by a rotation matrix. Of course it generates large files due to the current file format definition, but may still be useful to "non-Cartesian people" :-)

**Comment by zhengliuX** [comment]
> 2. adding 'first' and 'last' properties to the GRAD objects

This feature is significant for me. Now I calculated the first and last value in the interpreter, but the value is highly customizable in the Pulseq code. This could affect the trajectory if the developer doesn't know how the interpreter processes the first and last value, like force-set the first and last value to zero, or just calculate the value based on the gradient waveform.

Also **RF pulse extensions for parallel transmit** feature will definitely have a strong demand.

Zheng

**Reply by m-a-x-i-m-z** [reply]
yes, pTX is on our list, but it will take a while until we implement some extension. But we are happy to discuss what is needed and how it van be achieved. We can start a separate thread for that.

**Reply by zhengliuX** [reply]
Excellent. I would love to discuss about PTX.

**Reply by dabosch** [reply]
I started a bit of a discussion here:
https://github.com/pulseq/pulseq/discussions/69

**Comment by m-a-x-i-m-z** [comment]
Another extension that we've discussed once internally is a kind of a _soft delay_, which will allow the interpreter to modify  durations of the pure delay blocks (e.g. Pulseq blocks that contain no events).

**Reply by dabosch** [reply]
I think that can be highly interesting, allowing for e.g. the option to modify TE or TR on the console, if some additional labels and corresponding logic in the interpreter are added.

**Reply by zhengliuX** [reply]
I highly agree with the soft delay design. The ability of modifying TE and TR is really useful. I guess the key is to find a way to label the pure delay block,and the interpereter could synchronize the delay value on the UI, then get the value for executing the sequence. 

Also, is there any thoughts when the value of the soft delay is varying, like multiple TE.

Best,
Zheng

**Reply by m-a-x-i-m-z** [reply]
look [here](https://github.com/pulseq/pulseq/discussions/75#discussion-7617330)

**Comment by m-a-x-i-m-z** [comment]
So, I've just committed a first try to switch the file format, the internal storage in the event libraries and the write/read functions have been updated. There is a new function seq.write_v141() which can export the Pulseq file in the old(current) format. The new pulse definitions now look like:
```
# Format of RF events:
# id amplitude mag_id phase_id time_shape_id center delay freq phase use
# ..        Hz     ..       ..            ..     us    us   Hz   rad  ..
# Field 'use' is the initial of: excitation refocusing inversion saturation preparation other undefined
[RF]
1       129.71 1 2 3 4000 100 -424.504 10.6689 s
2       502.45 4 5 0 1000 130 -4500 28.2743 e
```
One can notice the letters 's' and 'e', which are really new. also the pulse center is now stored. 

Also for the ADC we have added the phase modulation vector, which refers to the common shapes library:
```
# Format of ADC events:
# id num dwell delay freq phase phase_id
# ..  ..    ns    us   Hz   rad       ..
[ADC]
1 192 3000 32 0 0 6
```

There is also a change in the arbitrary gradient storage, but it is less spectacular. 

*What however I wanted to discuss here is the following:* 
Currently no interpreter supports the new features. and as I've started to think of implementing such support, I've realized that the new features mentioned above can more or less be fully ignored by the interpreter. (This is why we actually have survived without them for so long.) So, with that I realize that these features have a character of extensions, which are not 100% for the interpreter to be able to play out the sequence. According to the current concept fixed in the Pulseq Specification document, interpreters are free to chose to ignore some or even all of the extensions. Based on that logic, it would be more advantageous to allow RF and ADC objects to have/define extensions (which would be the most essential change in the format towards v1.5.0) and actually to add things like ADC phase modulation or RF use to extensions. 
* The advantage of this approach would be that it would pave the way towards the further extensions, like those for pTX, RF shimming or ADC masking. 
* The disadvantage of the extension approach is the overhead of handling these features as extensions, e.g. in terms of coding in Matlab/Python, and also in terms of the performance. For instance the file writing/loading duration is substantially affected by using LABEL extension, that is reading/writing files without extensions would be faster (and eventually more robust). Furthermore, for the majority of Pulseq objects we actually adopt the same storage in memory, which basically means that the overhead of unpacking extensions will be imposed on every seq.getBlock() call, e.g. for plotting or k-space calculation and may eventually result is substantial slowdowns.

Your thoughts?

**Reply by dabosch** [reply]
I think adding extensions for the RF pulses potentially opens up a for more opportunities.
Right now, if you look at the gre.seq in the repo, [here](https://github.com/pulseq/pulseq/blob/8df1e293a469ef2cf277308459998a3007b0c3fb/tests/gre.seq#L344), there are 24 different RF pulses because of RF spoiling. The interpreter (or any other software) can not know if they are all the same pulse (e.g. in the same excitation pulse) or if they are different (e.g. different excitation pulses for 2 different contrasts in an MP2RAGE, or even a preparation pulse).

If we have extensions describing the pulses more, we could even give them IDs/Names. Suddenly, software could tell us "There are 2 excitation pulses in this sequence". We could then even make those pulses controllable from the interpreter (in siemens: use the FA field to scale the pulse flip angle).

This might be a bit far fetched, but would allow for quite some flexibility in the sequences.

**Reply by m-a-x-i-m-z** [reply]
Hi, 

good discussions, but actually things are slightly different: 

* the example with 24 pulses is a good one: currently the interpreter DOES know it is the same pulse because it shares the same shape. So currently the redundancy is actually very low. 
* the extension of the RF pulse will not address the problem of the proliferation of the numbers of pulses (although, as mentioned above, it is not a real problem), because for really addressing it we need a higher-lever object, like "RF pulse config", which will refer to an RF pulse and add some phase and frequency to it. This is because we would actually need a 1-to-N relation between the RF pulses and RF phase settings, which is not addressable by an extension of the RF pulse (this will result in 1-to-1 or N-to-1 topology). 

But in general I like the idea of Dario, it aligns well with what I've had in mind for the next-gen pTX & RF shimming. So I'd propose we save it to the next revision and for now in the interest of time stick to the current prototype implementation.

--M.

**Reply by zhengliuX** [reply]
One possible scenario. I added a RF amplitude factor in the interpreter, so people can multiply a factor to all the RF amplitudes. If we have this "use" field, we may can specify the factor to be used on specific usage precisely.

Best,
Zheng
